File: android/app/src/main/java/de/youspeed/android/alpha/BundleDownloadQueue.kt
package de.youspeed.android.alpha

/** Main-lane FIFO; closing a sheet does not cancel requested downloads. */
internal class BundleDownloadQueue<T>(private val id: (T) -> String) {
    private val entries = mutableListOf<T>()
    val ids: List<String> get() = entries.map(id)

    fun enqueue(entry: T, activeId: String?) {
        if (id(entry) != activeId && id(entry) !in ids) entries += entry
    }

    fun remove(entryId: String) { entries.removeAll { id(it) == entryId } }

    fun next(isBusy: Boolean): T? = if (isBusy || entries.isEmpty()) null else entries.removeAt(0)
}

File: iphone/SpeedConsumerApp/BundleDownloadQueue.swift
import Foundation

/// Serial, duplicate-free requests. Closing settings does not discard the queue.
struct BundleDownloadQueue<Entry: Identifiable> where Entry.ID == String {
    private(set) var entries: [Entry] = []
    var ids: [String] { entries.map(\.id) }

    mutating func enqueue(_ entry: Entry, activeID: String?) {
        guard entry.id != activeID, !ids.contains(entry.id) else { return }
        entries.append(entry)
    }

    mutating func remove(id: String) { entries.removeAll { $0.id == id } }

    mutating func next(isBusy: Bool) -> Entry? {
        guard !isBusy, !entries.isEmpty else { return nil }
        return entries.removeFirst()
    }
}
